### PR TITLE
[feature-wip](multi-catalog) Impl FileScanNode in be

### PR DESCRIPTION
**Pull request by SaintBacchus**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:
Define a new file scanner node for hms table in be.
This file scanner node is different from broker scan node as blow:
1. Broker scan node will define src slot and dest slot, there is two memory copy in it: first is from file to src slot and second from src to dest slot. Otherwise FileScanNode only have one stemp memory copy just from file to dest slot.
2. Broker scan node will read all the filed in the file to src slot and FileScanNode only read the need filed.
3. Broker scan node will convert type into string type for src slot and then use cast to convert to dest slot type, but FileScanNode will have the final type.

Now FileScanNode is a standalone code, but we will uniform the file scan and broker scan in the feature.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
5. Has unit tests been added: (Yes/No/No Need)
6. Has document been added or modified: (Yes/No/No Need)
7. Does it need to update dependencies: (Yes/No)
8. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/exec/file_arrow_scanner.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+#include "common/status.h"
+#include "exec/base_scanner.h"
+#include "vec/exec/file_scanner.h"
+#include "util/runtime_profile.h"
+
+namespace doris::vectorized {
+
+// VArrow scanner convert the data read from orc|parquet to doris's columns.
+class FileArrowScanner : public FileScanner {
```
Comment (morningman) [inline be/src/vec/exec/file_arrow_scanner.h:39]: What is different between `FileArrowScanner` and `VArrowScanner` ?
Looks like the `VArrowScanner` first read data into src block, and then convert to dest block?
And `FileArrowScanner` read data directly into dest block?

File: be/src/vec/exec/file_arrow_scanner.cpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        return Status::EndOfFile("EOF");
+    }
+    if (_batch == nullptr || _arrow_batch_cur_idx >= _batch->num_rows()) {
+        return _next_arrow_batch();
+    }
+    return status;
+}
+
+Status FileArrowScanner::get_next(vectorized::Block* block, bool* eof) {
+    std::cout << "HZW" << std::endl;
```
Comment (dujl) [inline be/src/vec/exec/file_arrow_scanner.cpp:136]: remove the unused cout

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalFileScanNode.java [hunk 3]
```diff
@@ -253,45 +264,40 @@ private void buildScanRange() throws UserException, IOException {
...
 
             TScanRangeLocations curLocations = newLocations(context.params);
             List<String> partitionValuesFromPath = BrokerUtil.parseColumnsFromPath(fileSplit.getPath().toString(),
                     partitionKeys);
-            int numberOfColumnsFromFile = context.slotDescByName.size() - partitionValuesFromPath.size();
 
-            TBrokerRangeDesc rangeDesc = createBrokerRangeDesc(fileSplit, partitionValuesFromPath,
-                    numberOfColumnsFromFile);
+            TFileRangeDesc rangeDesc = createFileRangeDesc(fileSplit, partitionValuesFromPath);
             rangeDesc.setHdfsParams(hdfsParams);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalFileScanNode.java:277]: Already call `setHdfsParams` in `createBrokerRangeDesc()`

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalFileScanNode.java [hunk 4]
```diff
@@ -253,45 +264,40 @@ private void buildScanRange() throws UserException, IOException {
...
 
-            curLocations.getScanRange().getBrokerScanRange().addToRanges(rangeDesc);
+            curLocations.getScanRange().getExtScanRange().getFileScanRange().addToRanges(rangeDesc);
             Log.debug("Assign to backend " + curLocations.getLocations().get(0).getBackendId()
                     + " with table split: " +  fileSplit.getPath()
                     + " ( " + fileSplit.getStart() + "," + fileSplit.getLength() + ")");
 
             // Put the last file
-            if (curLocations.getScanRange().getBrokerScanRange().isSetRanges()) {
+            if (curLocations.getScanRange().getExtScanRange().getFileScanRange().isSetRanges()) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalFileScanNode.java:285]: This `if` always return `true`?

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalFileScanNode.java [hunk 5]
```diff
@@ -253,45 +264,40 @@ private void buildScanRange() throws UserException, IOException {
...
-        List<String> partitionKeys = new ArrayList<>();
-        for (FieldSchema fieldSchema : hmsTable.getRemoteTable().getPartitionKeys()) {
-            partitionKeys.add(fieldSchema.getName());
-        }
+
 
         for (InputSplit split : inputSplits) {
             FileSplit fileSplit = (FileSplit) split;
 
             TScanRangeLocations curLocations = newLocations(context.params);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/ExternalFileScanNode.java:272]: Here you create `TScanRangeLocations` for every file?
That would be too many `TScanRangeLocations`.

File: fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java [hunk 6]
```diff
@@ -1686,49 +1688,55 @@ private PlanNode createScanNode(Analyzer analyzer, TableRef tblRef, SelectStmt s
...
-                scanNode = new HudiScanNode(ctx.getNextNodeId(), tblRef.getDesc(), "HudiScanNode",
-                        null, -1);
-                break;
-            default:
-                break;
-        }
-        if (scanNode instanceof OlapScanNode || scanNode instanceof EsScanNode || scanNode instanceof HiveScanNode) {
-            if (analyzer.enableInferPredicate()) {
-                PredicatePushDown.visitScanNode(scanNode, tblRef.getJoinOp(), analyzer);
+        if (Config.enable_multi_catalog) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/SingleNodePlanner.java:1691]: Add `HMS_EXTERNAL_TABLE` in `TableType`

File: fe/fe-core/src/main/java/org/apache/doris/common/Config.java [hunk 7]
```diff
@@ -1652,7 +1652,7 @@ public class Config extends ConfigBase {
      * Should be removed when this feature is ready.
      */
     @ConfField(mutable = false, masterOnly = true)
-    public static boolean enable_multi_catalog = false; // 1 min
+    public static boolean enable_multi_catalog = true; // 1 min
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1655]: set back to false.

File: be/src/vec/exec/file_arrow_scanner.cpp [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+    while (true) {
+        if (_next_range >= _ranges.size()) {
+            _scanner_eof = true;
+            return Status::OK();
+        }
+        const TFileRangeDesc& range = _ranges[_next_range++];
+        std::unique_ptr<FileReader> file_reader;
+
+        FileReader* hdfs_reader = nullptr;
```
Comment (dujl) [inline be/src/vec/exec/file_arrow_scanner.cpp:59]: reader could be hdfs_reader or s3_reader or other object store system, so use FileFactory::create_file_reader to create a file reader?

File: be/src/vec/exec/file_scan_node.cpp [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+
+    return Status::OK();
+}
+
+Status FileScanNode::start_scanners() {
+    {
+        std::unique_lock<std::mutex> l(_batch_queue_lock);
+        _num_running_scanners = 1;
+    }
+    _scanner_threads.emplace_back(&FileScanNode::scanner_worker, this, 0, _scan_ranges.size());
```
Comment (dujl) [inline be/src/vec/exec/file_scan_node.cpp:102]: always start only one thread to scan the ranges?

File: be/src/vec/exec/file_scan_node.cpp [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+        scan = new VFileParquetScanner(_runtime_state, runtime_profile(), scan_range.params,
+                                       scan_range.ranges, _pre_filter_texprs, counter);
+        break;
+    case TFileFormatType::FORMAT_ORC:
+        scan = new VFileORCScanner(_runtime_state, runtime_profile(), scan_range.params,
+                                   scan_range.ranges, _pre_filter_texprs, counter);
+        break;
+
+    default:
+        scan = new FileTextScanner(_runtime_state, runtime_profile(), scan_range.params,
```
Comment (dujl) [inline be/src/vec/exec/file_scan_node.cpp:317]: is json format handled by FileTextScanner?

File: be/src/vec/exec/file_arrow_scanner.cpp [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        }
+        // return error if not EOF
+        if (!status.is_end_of_file()) {
+            return status;
+        }
+        _cur_file_eof = true;
+        break;
+    }
+
+    return fill_block(block, eof);
```
Comment (dujl) [inline be/src/vec/exec/file_arrow_scanner.cpp:171]: fill_block do two works.
1. fill the data block
2. filter the block with pre filter expr
How about Separate filte_block from fill_block?

File: be/src/vec/exec/file_text_scanner.cpp [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+        }
+        size += 1;
+        // not first range will always skip one line
+        _skip_lines = 1;
+    }
+
+    // open line reader
+    switch (range.format_type) {
+    case TFileFormatType::FORMAT_CSV_PLAIN:
+        _cur_line_reader = new PlainTextLineReader(_profile, _cur_file_reader.get(), nullptr, size,
```
Comment (dujl) [inline be/src/vec/exec/file_text_scanner.cpp:203]: do we support compressed csv file?

File: be/src/vec/exec/file_text_scanner.cpp [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+    _next_range++;
+
+    return Status::OK();
+}
+
+Status FileTextScanner::_open_file_reader() {
+    const TFileRangeDesc& range = _ranges[_next_range];
+
+    FileReader* hdfs_reader = nullptr;
+    RETURN_IF_ERROR(HdfsReaderWriter::create_reader(range.hdfs_params, range.path,
```
Comment (dujl) [inline be/src/vec/exec/file_text_scanner.cpp:175]: do we support other file system such as s3?

File: gensrc/thrift/PlanNodes.thrift [hunk 14]
```diff
@@ -213,8 +215,54 @@ struct TEsScanRange {
...
+}
+
+struct TFileRangeDesc {
+    1: optional Types.TFileType file_type;
+    2: optional TFileFormatType format_type;
+    // Path of this range
+    3: optional string path;
+    // Offset of this file start
+    4: optional i64 start_offset;
+    // Size of this range, if size = -1, this means that will read to then end of file
```
Comment (dujl) [inline gensrc/thrift/PlanNodes.thrift:251]: then->the : read to the end of file

File: gensrc/thrift/PlanNodes.thrift [hunk 15]
```diff
@@ -213,8 +215,54 @@ struct TEsScanRange {
...
-struct TFileScanRange {
+struct TFileTextScanRangeParams {
+    3: optional i8 column_separator;
+    4: optional i8 line_delimiter;
+    // for multibytes separators
+    5: optional i32 column_separator_length = 1;
+    6: optional i32 line_delimiter_length = 1;
+    7: optional string column_separator_str;
+    8: optional string line_delimiter_str;
+}
```
Comment (dujl) [inline gensrc/thrift/PlanNodes.thrift:226]: does we support more options such as quote, escape, include header or not and so on?
suggest TFileTextScanRangeParams use map<k, v> struct, as that we could support more options without change the TFileTextScanRangeParams

Comment (morningman) [reply gensrc/thrift/PlanNodes.thrift:221]: `map` is not an efficient way for serde.
And if we add more options in future, we still need to modify BE or FE code.
So I think it is ok for now.

**Review by morningman — APPROVED** [review]
LGTM,
there are still some unresolved problem that will be fixed in other PRs.

**Pull request merged**